### PR TITLE
Add ability to hide comments

### PR DESCRIPTION
**Pull request by JoshLove-msft**
Adds a checkbox with label "Show comments" that appears in the upper right. It is checked by default. If unchecked, comments and code diagnostics are hidden. 

Also adds an icon next to code lines that have comments/diagnostics if the comments/diagnostics are hidden.

Also did a Format Document on Review.cshtml

### CODE REVIEW

**Comment by pakrym** [comment]
I like the feature but am not sure about the implementation, it doesn't inter-operate very well with existing commenting functionality, for example sometimes clicking on the element or the add button would open a commenting window and you would be able to comment without knowing if there are existed comments and sometimes comment window wouldn't even show up looking like a bug.

**Comment by JoshLove-msft** [comment]
> I like the feature but am not sure about the implementation, it doesn't inter-operate very well with existing commenting functionality, for example sometimes clicking on the element or the add button would open a commenting window and you would be able to comment without knowing if there are existed comments and sometimes comment window wouldn't even show up looking like a bug.
> 
> Non persistence across the revision switch is less of a concern and is easy to fix.

When would comment window not even show up when attempting to add a new comment (or maybe I am misinterpreting your comment)? I think it would always show up when clicking to add a new comment regardless of the state of the checkbox since it is a new element being added and the jquery selector only updates the elements that are already in the document when the checkbox is clicked.

Since it is checked by default, a user would have to intentionally uncheck it. However I can see them forgetting that they unchecked it and coming back later and adding the same comment again or something like that. I guess we could lock down adding new comments if they are currently being hidden?

**Comment by pakrym** [comment]
> When would comment window not even show up when attempting to add a new comment (or maybe I am misinterpreting your comment)?

Try hiding an existing thread and then clicking on add button for the same element. New element is added but under a hidden one.

>I guess we could lock down adding new comments if they are currently being hidden?

This would require people to unhide and find the element again to add a comment.


> Another idea would be to automatically toggle the comments back on when a user clicks to add a new comment. This would help prevent the duplicate comment concern.

Yeah, having some compact indication that there are comments/diagnostics but they are hidden, allow toggling them individually and having add show the existing comments would be the best option.

**Comment by JoshLove-msft** [comment]
> Try hiding an existing thread and then clicking on add button for the same element. New element is added but under a hidden one.

Sure, I thought you were saying that there were cases where a new comment couldn't be added.

> having some compact indication that there are comments/diagnostics but they are hidden

Do you mean per hidden comment thread? This seems to defeat the purpose of the feature, as the intent is to show solely the API without any clutter from comments.

> allow toggling them individually
 
This doesn't seem useful to me as we already have the ability to resolve comments. I don't see the use in hiding individual comments completely as this could lead to even more confusion, especially if we don't add the compact indication of the existence of comments on a thread by thread basis.

I do agree about showing the existing comments when attempting to add a new comment if they are currently hidden, but as for the other two features I guess we can chat more in person if you feel strongly about these.

**Comment by pakrym** [comment]
> Sure, I thought you were saying that there were cases where a new comment couldn't be added.

Yes, new comment can not be added to an existing thread.

> Do you mean per hidden comment thread? This seems to defeat the purpose of the feature, as the intent is to show solely the API without any clutter from comments.

It depends on how the indication is done, I'm sure we can think of something that doesn't clutter too much.

> This doesn't seem useful to me as we already have the ability to resolve comments. I don't see the use in hiding individual comments completely as this could lead to even more confusion, especially if we don't add the compact indication of the existence of comments on a thread by thread basis.

Imagine you're scrolling through APIs and want to leave a comment but would like to see existing ones first, scrolling to the top and toggling all threads on would be not very convenient.

**Comment by pakrym** [comment]
I though about something like:
image


It would represent that there are existing comments, and clicking on it would show them and allow commenting.

**Comment by JoshLove-msft** [comment]
> Imagine you're scrolling through APIs and want to leave a comment but would like to see existing ones first, scrolling to the top and toggling all threads on would be not very convenient.

I guess I was imagining this as a feature that would be used only when wanting to get an overall sense of the shape of the API. I didn't really think of it as something where users would want to actually interact with the tool while things are hidden, but I guess it is reasonable that they would want to have things hidden yet still be able to see individual comments for a specific API. This opens up the whole can of worms of managing the state of whether individual comments are hidden. I assume we would want to store off this information just like we do for resolved comments? Also, do you think the checkbox still makes sense given that we want to toggle individual comments? It is still useful to toggle things all off/on quickly, but the state of it is a bit weird when somethings are shown and some are hidden. I think that is probably okay.

**Comment by JoshLove-msft** [comment]
> I though about something like:
> image
> 
> It would represent that there are existing comments, and clicking on it would show them and allow commenting.

Yeah that looks pretty good, thanks.

File: src/dotnet/APIView/APIViewWeb/wwwroot/css/site.css [hunk 1]
```diff
@@ -422,4 +422,16 @@ code-inner {
...
+
+.icon-comments {
+    cursor: pointer;
+}
+
+.hidden {
+    visibility: hidden;
+}
+
+.inline {
```
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/wwwroot/css/site.css]: this style already exists

File: src/dotnet/APIView/APIViewWeb/wwwroot/css/site.css [hunk 2]
```diff
@@ -422,4 +422,16 @@ code-inner {
 
 .icon-package {
     background: url(/icons/package.png) center left no-repeat;
+}
+
+.icon-comments {
+    cursor: pointer;
+}
+
+.hidden {
```
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/wwwroot/css/site.css]: `.invisible`

File: src/dotnet/APIView/APIViewWeb/wwwroot/js/comments.ts [hunk 3]
```diff
@@ -109,4 +148,35 @@
...
+        }
+        else {
+            $(COMMENT_ROW_SEL).hide();
+            $(CODE_DIAGNOSTICS_SEL).hide();
+            $(ICON_COMMENTS_SEL).removeClass(HIDDEN)
+        }
+    }
+
+    function showSingleCommentAndDiagnostics(id) {
+        let $elem = getCommentsRow(id);
```
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/wwwroot/js/comments.ts]: remove variable

File: src/dotnet/APIView/APIViewWeb/wwwroot/js/comments.ts [hunk 4]
```diff
@@ -109,4 +148,35 @@
...
+        $elem.show();
+        getCodeRow(id).find(ICON_COMMENTS_SEL).addClass(HIDDEN);
+        getDiagnosticsRow(id).show();
+    }
+
+    function hideSingleComment(id) {
+        let $elem = getCommentsRow(id);
```
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/wwwroot/js/comments.ts]: remove var

File: src/dotnet/APIView/APIViewWeb/Pages/Assemblies/Review.cshtml [hunk 5]
```diff
@@ -41,6 +41,12 @@
                 <span class="dropdown-item"><b>@Model.ActiveConversations</b> active revision threads.</span>
                 <span class="dropdown-item"><b>@Model.TotalActiveConversations</b> total active threads.</span>
             </div>
+            <div class="checkbox">
```
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/Pages/Assemblies/Review.cshtml]: move this into the dropdown

File: src/dotnet/APIView/APIViewWeb/Pages/Assemblies/Review.cshtml [hunk 6]
```diff
@@ -49,49 +55,57 @@
...
-                            @if (lineDiagnostic.Text.StartsWith("DO"))
-                            {
-                            @:✅ <b>DO</b> @lineDiagnostic.Text.Substring(2)
-                            }
-                            else
+                            <a class="line-comment-button">+</a>
+                            @if (Model.Comments.TryGetThreadForLine(line.ElementId, out var _) || line.Diagnostics.Any())
                             {
-                            @lineDiagnostic.Text
+                                <span class="hidden icon icon-comments">💬</span>
```
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/Pages/Assemblies/Review.cshtml]: I wonder if we can keep this entirely client-side.

File: src/dotnet/APIView/APIViewWeb/Pages/Shared/_CommentThreadPartial.cshtml [hunk 7]
```diff
@@ -60,16 +60,17 @@
...
                         {
-                            <form method="post" asp-controller="Comments" asp-action="Unresolve" asp-route-reviewId="@Model.ReviewId">
+                            <form method="post" class="inline" asp-controller="Comments" asp-action="Unresolve" asp-route-reviewId="@Model.ReviewId">
                                 <button data-post-update="comments" type="submit" name="submit" value="Submit" class="btn btn-outline-secondary">Unresolve</button>
                             </form>
                         }
                         else
                         {
-                            <form method="post" asp-controller="Comments" asp-action="Resolve"  asp-route-reviewId="@Model.ReviewId">
+                            <form method="post" class="inline" asp-controller="Comments" asp-action="Resolve"  asp-route-reviewId="@Model.ReviewId">
```
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/Pages/Shared/_CommentThreadPartial.cshtml]: Why was this required?

File: src/dotnet/APIView/APIViewWeb/wwwroot/js/comments.ts [hunk 8]
```diff
@@ -109,4 +148,35 @@
         $(commentBox).replaceWith(partialViewResult);
         return false;
     }
+
+    function toggleAllCommentsAndDiagnosticsVisibility(element) {
+        if (element.checked) {
+            $(COMMENT_ROW_SEL).show();
```
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/wwwroot/js/comments.ts]: `.toggle(element.checked)` ?

File: src/dotnet/APIView/APIViewWeb/wwwroot/js/comments.ts [hunk 9]
```diff
@@ -109,4 +148,35 @@
...
+            $(CODE_DIAGNOSTICS_SEL).show();
+            $(ICON_COMMENTS_SEL).addClass(HIDDEN);
+        }
+        else {
+            $(COMMENT_ROW_SEL).hide();
+            $(CODE_DIAGNOSTICS_SEL).hide();
+            $(ICON_COMMENTS_SEL).removeClass(HIDDEN)
```
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/wwwroot/js/comments.ts]: toggleClass?

File: src/dotnet/APIView/APIViewWeb/wwwroot/js/comments.ts [hunk 10]
```diff
@@ -109,4 +148,35 @@
         $(commentBox).replaceWith(partialViewResult);
         return false;
     }
+
+    function toggleAllCommentsAndDiagnosticsVisibility(element) {
```
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/wwwroot/js/comments.ts]: I would take the value as a parameter here.

File: src/dotnet/APIView/APIViewWeb/Pages/Shared/_CommentThreadPartial.cshtml (hunk 7 above)
Comment (JoshLove-msft) [reply src/dotnet/APIView/APIViewWeb/Pages/Shared/_CommentThreadPartial.cshtml]: I added this so the buttons would sit next to each other.

File: src/dotnet/APIView/APIViewWeb/Pages/Assemblies/Review.cshtml (hunk 6 above)
Comment (JoshLove-msft) [reply src/dotnet/APIView/APIViewWeb/Pages/Assemblies/Review.cshtml]: I considered this but figured it was more performant to set on the server when we know there is a comment. I've added comments in the client side code that mentions where the icon is added in the server code.

File: src/dotnet/APIView/APIViewWeb/wwwroot/css/site.css [hunk 11]
```diff
@@ -418,4 +418,16 @@ code-inner {
...
+
+.icon-comments {
+    cursor: pointer;
+    position: absolute;
+    width: 16px;
+    height: 16px;
+    z-index: 1;
+}
+
+.inline {
```
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/wwwroot/css/site.css]: do you still need this?

File: src/dotnet/APIView/APIViewWeb/wwwroot/js/comments.ts [hunk 12]
```diff
@@ -122,9 +149,16 @@
             if (commentForm.length === 0) {
                 commentForm = $(createCommentForm()).insertAfter(reply);
             }
-
             reply.hide();
             commentForm.show();
+            commentsRow.show(); // ensure that entire comment row isn't being hidden
+        }
+
+        $(codeRow).next(SEL_CODE_DIAG).show(); // ensure that any code diagnostic for this row is shown in case it was previously hidden
```
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/wwwroot/js/comments.ts]: `getDiagnosticsRow`?

File: src/dotnet/APIView/APIViewWeb/wwwroot/js/comments.ts [hunk 13]
```diff
@@ -122,9 +149,16 @@
...
+
+        // If comment checkbox is unchecked, show the icon for new comment
+        if (!($("#show-comments-checkbox").get(0) as HTMLInputElement).checked) {
+            codeRow.find(SEL_COMMENT_ICON).removeClass(INVISIBLE);
```
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/wwwroot/js/comments.ts]: I would've turned this into `tooggleCommentIcon(id)` method.

File: src/dotnet/APIView/APIViewWeb/wwwroot/js/comments.ts [hunk 14]
```diff
@@ -122,9 +149,16 @@
...
-
             reply.hide();
             commentForm.show();
+            commentsRow.show(); // ensure that entire comment row isn't being hidden
+        }
+
+        $(codeRow).next(SEL_CODE_DIAG).show(); // ensure that any code diagnostic for this row is shown in case it was previously hidden
+
+        // If comment checkbox is unchecked, show the icon for new comment
+        if (!($("#show-comments-checkbox").get(0) as HTMLInputElement).checked) {
```
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/wwwroot/js/comments.ts]: `$("#show-comments-checkbox").val()`?

File: src/dotnet/APIView/APIViewWeb/wwwroot/js/comments.ts [hunk 15]
```diff
@@ -147,4 +181,27 @@
         $(commentBox).replaceWith(partialViewResult);
         return false;
     }
+
+    function toggleAllCommentsAndDiagnosticsVisibility(show: boolean) {
+        $(SEL_COMMENT_ROW).toggle(show);
```
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/wwwroot/js/comments.ts]: This can be rewriten as something like
```
$("[data-line-id]").foreach(function () {
	var lineId = getLineId(this);

	 getCommentsRow(id).toggle(show);
     getDiagnosticsRow(id).toggle(show);
	 tooggleCommentIcon(id, show);
});
```

And avoid all the complicated selector navigation.

File: src/dotnet/APIView/APIViewWeb/wwwroot/css/site.css (hunk 11 above)
Comment (JoshLove-msft) [reply src/dotnet/APIView/APIViewWeb/wwwroot/css/site.css]: Nope, removed

File: src/dotnet/APIView/APIViewWeb/wwwroot/js/comments.ts (hunk 12 above)
Comment (JoshLove-msft) [reply src/dotnet/APIView/APIViewWeb/wwwroot/js/comments.ts]: Updated

File: src/dotnet/APIView/APIViewWeb/wwwroot/js/comments.ts (hunk 14 above)
Comment (JoshLove-msft) [reply src/dotnet/APIView/APIViewWeb/wwwroot/js/comments.ts]: This seems to always return "on"

Comment (pakrym) [reply src/dotnet/APIView/APIViewWeb/wwwroot/js/comments.ts]: ah, it's `.prop('checked')`

File: src/dotnet/APIView/APIViewWeb/wwwroot/css/site.css (hunk 11 above)
Comment (pakrym) [reply src/dotnet/APIView/APIViewWeb/wwwroot/css/site.css]: Still shows

Comment (JoshLove-msft) [reply src/dotnet/APIView/APIViewWeb/wwwroot/css/site.css]: actually removed..

File: src/dotnet/APIView/APIViewWeb/wwwroot/js/comments.ts [hunk 16]
```diff
@@ -1,4 +1,14 @@
 ﻿$(() => {
+    let commentFormTemplate = $("#comment-form-template");
```
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/wwwroot/js/comments.ts]: rem.

File: src/dotnet/APIView/APIViewWeb/wwwroot/js/comments.ts [hunk 17]
```diff
@@ -1,4 +1,14 @@
...
+    const INVISIBLE = "invisible";
+    const SEL_CODE_DIAG = ".code-diagnostics";
+    const SEL_COMMENT_ROW = ".comment-row";
```
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/wwwroot/js/comments.ts]: rem

File: src/dotnet/APIView/APIViewWeb/wwwroot/js/comments.ts [hunk 18]
```diff
@@ -1,4 +1,14 @@
...
+    const SEL_COMMENT_ICON = ".icon-comments";
+    const SEL_COMMENT_CELL = ".comment-cell";
+    const SEL_CODE_LINE = ".code-line";
```
Comment (pakrym) [inline src/dotnet/APIView/APIViewWeb/wwwroot/js/comments.ts]: rem
